src/nghttp.cc

```diff
@@ -414,6 +482,14 @@ int submit_request(HttpClient *client, const Headers &headers, Request *req) {
 
   req->req_nva = std::move(build_headers);
 
+  if (expect_continue) {
+    auto timer = std::make_shared<ContinueTimer>(client->loop, req);
+    timer->start();
+
+    req->continue_timer = timer;
+    client->continue_timers.push_back(timer);
```

Comment:
After request was submitted successfully, when stream is closed (including the case when it is reset), on_stream_close_callback() is called.  If TCP connection closed, then on_stream_close_callback may not be called, but HttpClient::disconnect() will be called.
So you can stop timer in on_stream_close_callback().  And also you can iterate requests in HttpClient::disconnect(), and stop timer for each request.  Then we don't have to use std::vector, and std::shared_ptr/std::weak_ptr.
I think if we get final response, we should stop timer, and dispatch callback.  In that point, stream may have closed, but the error will be treated by nghttp2 library gracefully.

There is one more race condition when timer is fired, before HEADERS frame is actually sent.  This may happen when the number of  concurrent streams is very low, and outstanding requests takes long time.  To workaround this, instead of starting timer just after nghttp2_submit_headers(), do it in nghttp2_on_frame_send_callback().  This ensures that HEADERS frame has been dispatched.